### PR TITLE
Aditya/pruning

### PR DESCRIPTION
**Pull request by AdityaSripal**

### CODE REVIEW

**Comment by AdityaSripal** [comment]
Ready for initial review!!

TODO:

- [ ]  Run Benchmarks

- [ ] Write more tests

File: mutable_tree.go [hunk 1]
```diff
@@ -18,13 +18,20 @@ type MutableTree struct {
...
+// NewMutableTree returns a new tree with the specified cache size and datastore and pruning options
+// To maintain backwards compatibility, this function will initialize PruningStrategy{keepEvery: 1, keepRecent: 0}
 func NewMutableTree(db dbm.DB, cacheSize int) *MutableTree {
-	ndb := newNodeDB(db, cacheSize)
+	// memDB is initialized but should never be written to
+	memDB := dbm.NewMemDB()
+	return NewMutableTreePruningOpts(db, memDB, cacheSize, 1, 0)
+}
+
+func NewMutableTreePruningOpts(snapDB dbm.DB, recentDB dbm.DB, cacheSize int, keepEvery, keepRecent int64) *MutableTree {
```
Comment (fedekunze) [inline mutable_tree.go]: ditto godoc

File: mutable_tree.go [hunk 2]
```diff
@@ -427,6 +444,7 @@ func (tree *MutableTree) SaveVersion() ([]byte, int64, error) {
 // DeleteVersion deletes a tree version from disk. The version can then no
 // longer be accessed.
 func (tree *MutableTree) DeleteVersion(version int64) error {
+	debug("DELETE VERSION: %d\n", version)
```
Comment (fedekunze) [inline mutable_tree.go]:
```suggestion
```

File: mutable_tree.go [hunk 3]
```diff
@@ -552,8 +570,8 @@ func (tree *MutableTree) balance(node *Node) (newSelf *Node, orphaned []*Node) {
 
 func (tree *MutableTree) addOrphans(orphans []*Node) {
 	for _, node := range orphans {
-		if !node.persisted {
-			// We don't need to orphan nodes that were never persisted.
+		if !node.saved {
```
Comment (marbar3778) [inline mutable_tree.go]: is this if statement necessary?

File: basic_test.go [hunk 4]
```diff
@@ -185,12 +185,10 @@ func TestUnit(t *testing.T) {
...
 
 func TestRemove(t *testing.T) {
-	size := 10000
 	keyLen, dataLen := 16, 40
 
-	d := db.NewDB("test", "memdb", "")
-	defer d.Close()
-	t1 := NewMutableTree(d, size)
+	size := 10000
+	t1 := getTestTree(size)
```
[the remark this answers is not in the document]
Comment (AdityaSripal) [reply basic_test.go]: `size` is also used elsewhere in the test function. Easier to assign a variable and change its value then replace all constants if we want to ever make it something other than 1000

File: mutable_tree.go (hunk 2 above)
Comment (AdityaSripal) [reply mutable_tree.go]: There are debug statements in multiple parts of the code including `SaveVersion`. They will not print anything unless `debugging` is set to `true`

File: mutable_tree.go (hunk 3 above)
Comment (AdityaSripal) [reply mutable_tree.go]: I'm not sure it is even possible to enter that conditional but i just replicated the previous behavior. I would feel more comfortable taking it out after i've fully finished testing. Will leave unresolved for now.

**Comment by marbar3778** [comment]
want to add the benchmarks into a file so we can keep track?

File: PRUNING.md [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+Setting Pruning fields in the IAVL tree can optimize performance by only writing versions to disk if they are meant to be persisted indefinitely. Versions that are known to be deleted eventually are temporarily held in memory until they are ready to be pruned. This greatly reduces the I/O load of IAVL.
+
+We can set custom pruning fields in IAVL using: `NewMutableTreePruningOpts`
+
+
+## Current design
+
+### NodeDB
+NodeDB has extra fields:
+```
```
Comment (fedekunze) [inline PRUNING.md]:
```suggestion

```go
```

File: PRUNING.md [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+// Pruning fields
+keepEvery  int64n // Saves version to disk periodically
+keepRecent int64  // Saves recent versions in memory
+```
+
+If version is not going to be persisted to disk, the version is simply saved in `recentDB` (typically a `memDB`)
+If version is persisted to disk, the version is written to `recentDB` **and** `snapshotDB` (typically `levelDB`)
+
+#### Orphans:
+Save orphan to memDB under `o|toVersion|fromVersion`.
```
Comment (fedekunze) [inline PRUNING.md]:
```suggestion

Save orphan to `memDB` under `o|toVersion|fromVersion`.
```

File: benchmarks/prune_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		}
+	}()
+
+	runtime.GC()
+
+	// always initialize tree with goleveldb as snapshotDB and memDB as recentDB
+	snapDB := db.NewDB("test", "goleveldb", dirName)
+	defer snapDB.Close()
+
+	// var mem runtime.MemStats
```
Comment (fedekunze) [inline benchmarks/prune_test.go]: leftover comments

File: benchmarks/prune_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+		// create 5 keys per version
+		for j := 0; j < 5; j++ {
+			t.Set(randBytes(keyLen), randBytes(dataLen))
+		}
+		_, _, err := t.SaveVersion()
+		if err != nil {
+			b.Errorf("Can't save version %d: %v", i, err)
+		}
+		// // Pause timer to garbage-collect and remeasure memory usage
+		// b.StopTimer()
```
Comment (fedekunze) [inline benchmarks/prune_test.go]: ditto

File: benchmarks/prune_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	//fmt.Printf("Maxmimum Memory usage was %0.2f MB at height %d\n", float64(memSize)/1000000, maxVersion)
+	b.StopTimer()
+}
+
+func BenchmarkPruningStrategies(b *testing.B) {
+	ps := []pruningstrat{
+		{1, 0},   // default pruning strategy
+		{0, 1},   // keep single recent version
+		{100, 5}, // simple pruning
+		// {1000, 10},   // average pruning
```
Comment (fedekunze) [inline benchmarks/prune_test.go]: ditto

File: mutable_tree.go [hunk 10]
```diff
@@ -200,10 +208,12 @@ func (tree *MutableTree) recursiveRemove(node *Node, key []byte) ([]byte, *Node,
 
 		if len(orphaned) == 0 {
 			return node.hash, node, nil, value, orphaned
-		} else if newLeftHash == nil && newLeftNode == nil { // left node held value, was removed
-			return node.rightHash, node.rightNode, node.key, value, orphaned
 		}
 		orphaned = append(orphaned, node)
+		if newLeftHash == nil && newLeftNode == nil { // left node held value, was removed
+			orphaned = append(orphaned, node)
```
Comment (fedekunze) [inline mutable_tree.go]: Is this line duplicated?

Comment (fedekunze) [reply mutable_tree.go]: It seems it is from the right node below
```suggestion
```

File: options.go [hunk 11]
```diff
@@ -0,0 +1,36 @@
+package iavl
+
+// Define Options for customizing pruning/writing strategy of IAVL tree
```
Comment (fedekunze) [inline options.go]:
```suggestion
// Options define customized pruning/writing strategies for the IAVL tree
```

File: options.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+type Options struct {
+	KeepEvery  int64
+	KeepRecent int64
+	Sync       bool
+}
+
+// Returns the default options for IAVL
```
Comment (fedekunze) [inline options.go]:
```suggestion
// DefaultOptions returns the default options for IAVL
```

File: options.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+func DefaultOptions() *Options {
+	return &Options{
+		KeepEvery:  1,
+		KeepRecent: 0,
+		Sync:       false,
+	}
+}
+
+// Return Options with given pruning strategy. Sync=false
```
Comment (fedekunze) [inline options.go]:
```suggestion
// PruningOptions return Options with a given pruning strategy. Sync is set to false by default.
```

File: options.go [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+func PruningOptions(keepEvery, keepRecent int64) *Options {
+	return &Options{
+		KeepEvery:  keepEvery,
+		KeepRecent: keepRecent,
+		Sync:       false,
+	}
+}
+
+// Return Options intended for benchmark tests
+// with given pruning strategy. Sync = true
```
Comment (fedekunze) [inline options.go]:
```suggestion
// with a given pruning strategy. Sync is set to true by default.
```

File: options.go [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+// Return Options with given pruning strategy. Sync=false
+func PruningOptions(keepEvery, keepRecent int64) *Options {
+	return &Options{
+		KeepEvery:  keepEvery,
+		KeepRecent: keepRecent,
+		Sync:       false,
+	}
+}
+
+// Return Options intended for benchmark tests
```
Comment (fedekunze) [inline options.go]:
```suggestion
// BenchingOptions returns Options intended for benchmark tests
```

File: benchmarks/prune_test.go (hunk 7 above)
Comment (AdityaSripal) [reply benchmarks/prune_test.go]: currently in process of getting this to work correctly. will remove once i've fixed this

File: mutable_tree.go (hunk 10 above)
Comment (AdityaSripal) [reply mutable_tree.go]: correct! thanks for the catch

File: pruning_test.go [hunk 16]
```diff
@@ -0,0 +1,462 @@
...
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"os"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	db "github.com/tendermint/tm-db")
```
Comment (golangcibot) [inline pruning_test.go]: File is not `gofmt`-ed with `-s` (from `gofmt`)
```suggestion
	db "github.com/tendermint/tm-db"
)
```

File: pruning_test.go [hunk 17]
```diff
@@ -0,0 +1,463 @@
...
+	require.Nil(t, err, "Error in getRoots")
+	numRoots := 1000/keepEvery + keepRecent
+	// decrement if there is overlap between snapshot and recent versions
+	if 1000%keepEvery == 0 {
+		numRoots--
+	}
+	require.Equal(t, numRoots, int64(len(roots)), "nodeDB does not contain expected number of roots")
+}
+
+func TestDeleteOrphans(t *testing.T) {
```
Comment (golangcibot) [inline pruning_test.go]: Function 'TestDeleteOrphans' is too long (73 > 60) (from `funlen`)

File: nodedb.go [hunk 18]
```diff
@@ -324,18 +499,41 @@ func (ndb *nodeDB) cacheNode(node *Node) {
...
+	}
+	ndb.snapshotBatch = ndb.snapshotDB.NewBatch()
+	ndb.recentBatch = ndb.recentDB.NewBatch()
 }
 
 func (ndb *nodeDB) getRoot(version int64) []byte {
-	return ndb.db.Get(ndb.rootKey(version))
+	if ndb.isRecentVersion(version) {
+		memroot := ndb.recentDB.Get(ndb.rootKey(version))
+		// TODO: maybe I shouldn't check in snapshot if it isn't here
```
Comment (golangcibot) [inline nodedb.go]: nodedb.go:530: Line contains TODO/BUG/FIXME: "TODO: maybe I shouldn't check in snapsho..." (from `godox`)

File: CHANGELOG_PENDING.md [hunk 19]
```diff
@@ -6,4 +6,10 @@ Special thanks to external contributors on this release:
 
 ### BREAKING CHANGES
 
+- [/#158] NodeDB constructor must provide `keepRecent` and `keepEvery` fields to define PruningStrategy. All Save functionality must specify whether they should flushToDisk as well using `flushToDisk` boolean argument. All Delete functionality must specify whether object should be deleted from memory only using the `memOnly` boolean argument.
+
 ### IMPROVEMENTS
+
+- [/#46](https://github.com/tendermint/iavl/issues/46) Removal of all instances of cmn from tendermint in Ival repo
```
Comment (Liamsi) [inline CHANGELOG_PENDING.md]: 🎉

File: Makefile [hunk 20]
```diff
@@ -49,6 +49,9 @@ fullbench:
 		go test $(LDFLAGS) -bench=Mem . && \
 		go test $(LDFLAGS) -timeout=60m -bench=LevelDB .
 
+benchprune:
+	cd benchmarks && \
+		go test -bench=PruningStrategies -timeout=24h
```
Comment (Liamsi) [inline Makefile]: Maybe add a line somewhere how to reproduce the benchmarks?

File: benchmarks/results/158-pruning/sdk-pruning-1-0-results.txt [hunk 21]
```diff
@@ -0,0 +1,164 @@
+Starting SimulateFromSeed with randomness created with seed 53
+Randomized simulation params:
```
Comment (Liamsi) [inline benchmarks/results/158-pruning/sdk-pruning-1-0-results.txt]: Running the sdk simul is probably important and the right thing to do. Not sure these files should go into the iavl repo though. It would be good to summarize this somewhere. And when the sdk catches up with the iavl version, the sdk team can add the benchmarks (or a condensed summary) in the sdk repo (or sth short in the changelog and a description how to reproduce).

Comment (marbar3778) [reply benchmarks/results/158-pruning/sdk-pruning-1-0-results.txt]: I believe @AdityaSripal and I briefly decided that if there is a PR that is claiming improvements then benchmarks should be added, then upon doing a release a final benchmark will be created prior to release and saved in the repo while the others are deleted. I will document this

File: logger.go [hunk 22]
```diff
@@ -4,8 +4,20 @@ import (
...
+)
+
 func debug(format string, args ...interface{}) {
-	if false {
+	if debugging {
 		fmt.Printf(format, args...)
 	}
 }
+
+func startDebugger() {
```
Comment (golangcibot) [inline logger.go]: `startDebugger` is unused (from `deadcode`)

File: logger.go [hunk 23]
```diff
@@ -4,8 +4,20 @@ import (
...
+	debugging = true
+}
+
+func endDebugger() {
```
Comment (golangcibot) [inline logger.go]: `endDebugger` is unused (from `deadcode`)

File: pruning_test.go [hunk 24]
```diff
@@ -0,0 +1,463 @@
...
+
+	keepRecent := int64(1) //keep 1 version in memDB
+	keepEvery := int64(10)
+	mt := NewMutableTreeWithOpts(db, mdb, 5, PruningOptions(keepEvery, keepRecent))
+
+	for v := 0; v < 10; v++ {
+		for i := 0; i < 10; i++ {
+			mt.Set([]byte(fmt.Sprintf("v%d:%d", v, i)), []byte(fmt.Sprintf("Val:v:%d:%d", v, i)))
+		}
+		mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: Error return value of `mt.SaveVersion` is not checked (from `errcheck`)

File: pruning_test.go [hunk 25]
```diff
@@ -0,0 +1,463 @@
...
+
+	numNodes := mt.nodeSize()
+
+	for v := 0; v < 10; v++ {
+		for i := 0; i < 10; i++ {
+			key := fmt.Sprintf("v%d:%d", v, i)
+			_, removed := mt.Remove([]byte(key))
+			require.True(t, removed, "Key %s could not be removed", key)
+		}
+		mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: Error return value of `mt.SaveVersion` is not checked (from `errcheck`)

File: pruning_test.go [hunk 26]
```diff
@@ -0,0 +1,463 @@
...
+	}
+
+	require.Equal(t, numNodes, len(mt.ndb.nodesFromDB(mt.ndb.snapshotDB)), "Number of Nodes in snapshotDB are unexpected")
+
+	// Delete only non-empty tree in snapshotDB
+	mt.DeleteVersion(10)
```
Comment (golangcibot) [inline pruning_test.go]: Error return value of `mt.DeleteVersion` is not checked (from `errcheck`)

File: pruning_test.go [hunk 27]
```diff
@@ -0,0 +1,463 @@
...
+	require.Equal(t, 0, size, "Not all orphans deleted")
+
+	size = 0
+	traverseOrphansFromDB(mt.ndb.recentDB, fn)
+	require.Equal(t, 0, size, "Not all orphans deleted from RecentDB")
+
+	require.Equal(t, len(mt.ndb.nodesFromDB(mt.ndb.snapshotDB)), len(mt.ndb.nodesFromDB(mt.ndb.recentDB)), "DB sizes should be the same")
+
+	for i := 1; i < 5; i++ {
+		mt.DeleteVersion(int64(i))
```
Comment (golangcibot) [inline pruning_test.go]: Error return value of `mt.DeleteVersion` is not checked (from `errcheck`)

File: pruning_test.go [hunk 28]
```diff
@@ -0,0 +1,463 @@
...
+		if n.version <= 100 {
+			numSnapNodes--
+		}
+		return false
+	}
+	mt.root.traverse(mt.ImmutableTree, true, fn)
+
+	require.Equal(t, numSnapNodes, len(mt.ndb.nodesFromDB(mt.ndb.snapshotDB)))
+
+	mt.DeleteVersion(100)
```
Comment (golangcibot) [inline pruning_test.go]: Error return value of `mt.DeleteVersion` is not checked (from `errcheck`)

File: pruning_test.go [hunk 29]
```diff
@@ -0,0 +1,463 @@
...
+
+		// check that root exists in nodeDB
+		lv, err := mt.LazyLoadVersion(ver)
+		require.Equal(t, ver, lv, "Version returned by LazyLoadVersion is wrong")
+		require.Nil(t, err, "Version should exist in nodeDB")
+	}
+
+	// check all expected versions are available.
+	for j := keepEvery; j <= mt.Version(); j += keepEvery {
+		require.True(t, mt.VersionExists(int64(j)), "Expected snapshot version: %d to be available in nodeDB. KeepEvery: %d, KeepRecent: %d", j, PruningOptions(keepEvery, keepRecent))
```
Comment (golangcibot) [inline pruning_test.go]: unnecessary conversion (from `unconvert`)

File: pruning_test.go [hunk 30]
```diff
@@ -0,0 +1,463 @@
...
+	}
+	for k := mt.Version() - keepRecent + 1; k <= mt.Version(); k++ {
+		require.True(t, mt.VersionExists(int64(k)), "Expected recent version: %d to be available in nodeDB. KeepEvery: %d, KeepRecent: %d", k, PruningOptions(keepEvery, keepRecent))
```
Comment (golangcibot) [inline pruning_test.go]: unnecessary conversion (from `unconvert`)

File: mutable_tree.go (hunk 2 above)
Comment (tnachen) [reply mutable_tree.go]: Should we use a logging library here? we can then set logging level

Comment (marbar3778) [reply mutable_tree.go]: can you make an issue with this, we shouldn't expand the scope of this PR into logging as well

Comment (tnachen) [reply mutable_tree.go]: Added #192

**Comment by zmanian** [comment]
@tnachen is taking this over I believe.

**Comment by tnachen** [comment]
Yes syncing with Aditya today about this patch

**Comment by tnachen** [comment]
I just did some benchmarking with the Cosmos SDK test-sim-benchmark with different configurations, and measure the benchmark completion time, and also looking into CPU / mem usage. It's not an exhaustive benchmark but overall can see the big improvements with pruning in terms of performance.

https://docs.google.com/spreadsheets/d/1p9eZ4LIq5wSjogb03_l59AzOMK_Gfu8ncWT6Ujua4nU/edit?usp=sharing

Overall CPU processing time and memory doesn't seem to differ that much regardless what recent/every setting chosen. The biggest different is time (compaction, etc).

File: mutable_tree.go [hunk 31]
```diff
@@ -18,13 +18,41 @@ type MutableTree struct {
...
+// To maintain backwards compatibility, this function will initialize PruningStrategy{keepEvery: 1, keepRecent: 0}
+func NewMutableTree(db dbm.DB, cacheSize int) (*MutableTree, error) {
+	// memDB is initialized but should never be written to
+	memDB := dbm.NewMemDB()
+	return NewMutableTreeWithOpts(db, memDB, cacheSize, nil)
+}
+
+func validateOptions(opts *Options) error {
+	switch {
+		case opts == nil:
```
Comment (golangcibot) [inline mutable_tree.go]: File is not `goimports`-ed (from `goimports`)

File: pruning_test.go [hunk 32]
```diff
@@ -0,0 +1,491 @@
...
+
+	require.Equal(t, mt.nodeSize(), len(mt.ndb.nodesFromDB(mt.ndb.snapshotDB)))
+}
+
+/* Test Pruning Edge Cases */
+func TestNoSnapshots(t *testing.T) {
+	db, mdb, close := getTestDBs()
+	defer close()
+
+	keepRecent := rand.Int63n(8) + 2                                        //keep at least 2 versions in memDB
```
Comment (golangcibot) [inline pruning_test.go]: File is not `goimports`-ed (from `goimports`)
```suggestion
	keepRecent := rand.Int63n(8) + 2                                             //keep at least 2 versions in memDB
```

File: testutils_test.go [hunk 33]
```diff
@@ -4,6 +4,7 @@ package iavl
 import (
 	"bytes"
 	"fmt"
+	"github.com/stretchr/testify/require"
```
Comment (golangcibot) [inline testutils_test.go]: File is not `goimports`-ed (from `goimports`)
```suggestion

```

File: pruning_test.go [hunk 34]
```diff
@@ -0,0 +1,491 @@
...
+	for i := 0; i < 1000; i++ {
+		// set 5 keys per version
+		for j := 0; j < 5; j++ {
+			key := make([]byte, 8)
+			val := make([]byte, 8)
+			binary.BigEndian.PutUint64(key, uint64(rand.Int63()))
+			binary.BigEndian.PutUint64(val, uint64(rand.Int63()))
+			mt.Set(key, val)
+		}
+		_, _, err := mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: shadow: declaration of "err" shadows declaration at line 31 (from `govet`)

File: pruning_test.go [hunk 35]
```diff
@@ -0,0 +1,491 @@
...
+	versions := mt.AvailableVersions()
+	// check that all available versions are expected.
+	for _, v := range versions {
+		ver := int64(v)
+		// check that version is supposed to exist given pruning strategy
+		require.True(t, ver%keepEvery == 0 || mt.Version()-ver <= keepRecent,
+			"Version: %d should not exist. KeepEvery: %d, KeepRecent: %d", v, PruningOptions(keepEvery, keepRecent))
+
+		// check that root exists in nodeDB
+		lv, err := mt.LazyLoadVersion(ver)
```
Comment (golangcibot) [inline pruning_test.go]: shadow: declaration of "err" shadows declaration at line 31 (from `govet`)

File: pruning_test.go [hunk 36]
```diff
@@ -0,0 +1,491 @@
...
+	keepRecent := int64(1) //keep 1 version in memDB
+	keepEvery := int64(10)
+	mt, err := NewMutableTreeWithOpts(db, mdb, 5, PruningOptions(keepEvery, keepRecent))
+	require.NoError(t, err)
+
+	for v := 0; v < 10; v++ {
+		for i := 0; i < 10; i++ {
+			mt.Set([]byte(fmt.Sprintf("v%d:%d", v, i)), []byte(fmt.Sprintf("Val:v:%d:%d", v, i)))
+		}
+		_, _, err := mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: shadow: declaration of "err" shadows declaration at line 201 (from `govet`)

File: pruning_test.go [hunk 37]
```diff
@@ -0,0 +1,491 @@
...
+
+	numNodes := mt.nodeSize()
+
+	for v := 0; v < 10; v++ {
+		for i := 0; i < 10; i++ {
+			key := fmt.Sprintf("v%d:%d", v, i)
+			_, removed := mt.Remove([]byte(key))
+			require.True(t, removed, "Key %s could not be removed", key)
+		}
+		_, _, err := mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: shadow: declaration of "err" shadows declaration at line 201 (from `govet`)

File: pruning_test.go [hunk 38]
```diff
@@ -0,0 +1,491 @@
...
+	numSnapNodes := 0
+	for i := 0; i < 200; i++ {
+		// set 5 keys per version
+		var key, val []byte
+		for j := 0; j < 5; j++ {
+			key = []byte(fmt.Sprintf("%d: Key:v%d:i%d", rand.Int63(), i+1, j))
+			val = []byte(fmt.Sprintf("Val:v%d:i%d", i, j))
+		}
+		mt.Set(key, val)
+		_, _, err := mt.SaveVersion()
```
Comment (golangcibot) [inline pruning_test.go]: shadow: declaration of "err" shadows declaration at line 362 (from `govet`)

File: PRUNING.md [hunk 39]
```diff
@@ -0,0 +1,38 @@
+# Pruning
```
Comment (marbar3778) [inline PRUNING.md]: should this file be moved to the docs section?

**Comment by marbar3778** [comment]
@tnachen is there a recommended config for users? We should document this.

File: benchmarks/prune_test.go [hunk 40]
```diff
@@ -56,37 +56,36 @@ func runBlockChain(b *testing.B, prefix string, keepEvery int64, keepRecent int6
...
 		runtime.GC()
+		runtime.ReadMemStats(&mem)
+		// update memSize if it has increased after saveVersion
+		if memSize < mem.Alloc {
+			memSize = mem.Alloc
+			maxVersion = i
+		}
 		b.StartTimer()
 	}
+
```
Comment (AdityaSripal) [inline benchmarks/prune_test.go]: Add comment to pause timer so we don't measure db-closing time

Also, remove print statement comment

File: benchmarks/prune_test.go [hunk 41]
```diff
@@ -56,37 +56,36 @@ func runBlockChain(b *testing.B, prefix string, keepEvery int64, keepRecent int6
...
-		// b.StartTimer()
+
+		// Pause timer to garbage-collect and remeasure memory usage
 		b.StopTimer()
 		runtime.GC()
+		runtime.ReadMemStats(&mem)
+		// update memSize if it has increased after saveVersion
+		if memSize < mem.Alloc {
+			memSize = mem.Alloc
+			maxVersion = i
```
Comment (marbar3778) [inline benchmarks/prune_test.go]: maxVersion is undefined here

File: mutable_tree_test.go [hunk 42]
```diff
@@ -1,15 +1,76 @@
...
+	_, _, err = tree.SaveVersion()
+	require.NoError(t, err)
+
+	require.NoError(t, tree.DeleteVersion(version))
+
+	k1Value, _, _ := tree.GetVersionedWithProof([]byte("k1"), version)
+	require.Nil(t, k1Value)
+
+	key := tree.ndb.rootKey(version)
+	memDb.Set(key, hash)
```
Comment (golangcibot) [inline mutable_tree_test.go]: Error return value of `memDb.Set` is not checked (from `errcheck`)

File: nodedb.go [hunk 43]
```diff
@@ -179,55 +242,107 @@ func (ndb *nodeDB) SaveOrphans(version int64, orphans map[string]int64) {
...
-			debug("DELETE predecessor:%v fromVersion:%v toVersion:%v %X\n", predecessor, fromVersion, toVersion, hash)
-			ndb.batch.Delete(ndb.nodeKey(hash))
+func (ndb *nodeDB) deleteOrphans(version int64, memOnly bool) {
+	if ndb.opts.KeepRecent != 0 {
+		ndb.deleteOrphansMem(version)
+	}
+	if ndb.isSnapshotVersion(version) && !memOnly {
+		predecessor := getPreviousVersionFromDB(version, ndb.snapshotDB)
+		traverseOrphansVersionFromDB(ndb.snapshotDB, version, func(key, hash []byte) {
+			ndb.snapshotDB.Delete(key)
```
Comment (golangcibot) [inline nodedb.go]: Error return value of `ndb.snapshotDB.Delete` is not checked (from `errcheck`)

File: nodedb.go [hunk 44]
```diff
@@ -342,18 +534,44 @@ func (ndb *nodeDB) cacheNode(node *Node) {
...
 func (ndb *nodeDB) Commit() {
 	ndb.mtx.Lock()
 	defer ndb.mtx.Unlock()
 
-	ndb.batch.Write()
-	ndb.batch.Close()
-	ndb.batch = ndb.db.NewBatch()
+	if ndb.opts.KeepEvery != 0 {
+		if ndb.opts.Sync {
+			ndb.snapshotBatch.WriteSync()
```
Comment (golangcibot) [inline nodedb.go]: Error return value of `ndb.snapshotBatch.WriteSync` is not checked (from `errcheck`)

File: nodedb.go [hunk 45]
```diff
@@ -342,18 +534,44 @@ func (ndb *nodeDB) cacheNode(node *Node) {
...
+		} else {
+			ndb.snapshotBatch.Write()
```
Comment (golangcibot) [inline nodedb.go]: Error return value of `ndb.snapshotBatch.Write` is not checked (from `errcheck`)

File: nodedb.go [hunk 46]
```diff
@@ -342,18 +534,44 @@ func (ndb *nodeDB) cacheNode(node *Node) {
...
+		}
+		ndb.snapshotBatch.Close()
+	}
+	if ndb.opts.KeepRecent != 0 {
+		if ndb.opts.Sync {
+			ndb.recentBatch.WriteSync()
```
Comment (golangcibot) [inline nodedb.go]: Error return value of `ndb.recentBatch.WriteSync` is not checked (from `errcheck`)

File: nodedb.go [hunk 47]
```diff
@@ -342,18 +534,44 @@ func (ndb *nodeDB) cacheNode(node *Node) {
...
+		} else {
+			ndb.recentBatch.Write()
```
Comment (golangcibot) [inline nodedb.go]: Error return value of `ndb.recentBatch.Write` is not checked (from `errcheck`)
